rustworkx-core/src/community/louvain.rs

```diff
@@ -0,0 +1,432 @@
+            v[s].insert(node);
+        }
+        v
+    }
+}
+
+/// Performs one level of the Louvain algorithm.
+///
+/// Arguments:
+///
+/// * `partition`: The current partition of the input graph
+/// * `m`: Total weight of the edges of `graph`
+/// * `resolution` : controls whether the algorithm favors larger communities (`resolution < 1`) or smaller communities (`resolution < 1`)
+/// * `gain_threshold` : minimum acceptable increase in modularity
+/// * `seed` : optional seed to determine the order in which we consider moving each node into a neighboring community
+///
+/// Returns true if it was possible to meet the specified `gain_threshold` by combining nodes into communities.
+fn one_level_undirected<G>(
+    partition: &mut Partition<G>,
+    m: f64,
+    resolution: f64,
+    gain_threshold: f64,
+    seed: Option<u64>,
+) -> bool
+where
+    G: Modularity,
+{
+    let inner_graph = partition.to_inner_graph();
+
+    let node_count = inner_graph.node_count();
+
+    let mut degrees = inner_graph.degrees();
+    let mut s_tot = degrees.clone();
+
+    // Start by placing each node into its own community
+    let mut node_to_community: Vec<usize> = (0..node_count).collect();
+    let mut total_gain = 0.0;
+    loop {
+        let mut performed_move = false;
+
+        let mut node_shuffle: Pcg32 = match seed {
+            Some(rng_seed) => Pcg32::seed_from_u64(rng_seed),
+            None => Pcg32::from_entropy(),
+        };
+
+        // Try moving each node into a neighboring community, in the order
+        // determined by `seed`. For each node, select the neighboring community
+        // that gives the largest increase in modularity (if any).
+        for node in rand::seq::index::sample(&mut node_shuffle, node_count, node_count) {
+            let neighbor_weights = inner_graph.neighbor_community_weights(node, &node_to_community);
+
+            let mut best_gain = 0.0;
+            let init_com = node_to_community[node];
+            let deg = degrees[init_com];
+            let mut best_com = init_com;
+            let two_m_sq = 2.0 * m * m;
+
+            degrees[best_com] -= deg;
+
+            let delta = if let Some(&w) = neighbor_weights.get(&best_com) {
+                w
+            } else {
+                0.0
+            };
+            let remove_cost = -delta / m + resolution * (s_tot[best_com] * deg) / two_m_sq;
+
+            for (&nbr_com, &wt) in neighbor_weights.iter() {
+                let gain = remove_cost + wt / m - resolution * s_tot[nbr_com] * deg / two_m_sq;
+                if gain > best_gain {
+                    best_gain = gain;
+                    best_com = nbr_com;
+                }
+            }
+
+            s_tot[best_com] += deg;
+
+            if best_com != init_com {
+                performed_move = true;
+                total_gain += best_gain;
+                node_to_community[node] = best_com;
+            }
+        }
+        if !performed_move {
+            break;
```

Comment:
I think we should loop on `while performed_move` or something like that to make the intention clearer.

Reply:
Done in 234659e.